## Rubric

Instructions: DELETE this cell before you submit via a `git push` to your repo before deadline. This cell is for your reference only and is not needed in your report. 

Scoring: Out of 10 points

- Each Developing  => -2 pts
- Each Unsatisfactory/Missing => -4 pts
  - until the score is 

If students address the detailed feedback in a future checkpoint they will earn these points back


|                  | Unsatisfactory                                                                                                                                                                                                    | Developing                                                                                                                                                                                              | Proficient                                     | Excellent                                                                                                                              |
|------------------|-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|------------------------------------------------|----------------------------------------------------------------------------------------------------------------------------------------|
| Data relevance   | Did not have data relevant to their question. Or the datasets don't work together because there is no way to line them up against each other. If there are multiple datasets, most of them have this trouble | Data was only tangentially relevant to the question or a bad proxy for the question. If there are multiple datasets, some of them may be irrelevant or can't be easily combined.                       | All data sources are relevant to the question. | Multiple data sources for each aspect of the project. It's clear how the data supports the needs of the project.                         |
| Data description | Dataset or its cleaning procedures are not described. If there are multiple datasets, most have this trouble                                                                                              | Data was not fully described. If there are multiple datasets, some of them are not fully described                                                                                                      | Data was fully described                       | The details of the data descriptions and perhaps some very basic EDA also make it clear how the data supports the needs of the project. |
| Data wrangling   | Did not obtain data. They did not clean/tidy the data they obtained.  If there are multiple datasets, most have this trouble                                                                                 | Data was partially cleaned or tidied. Perhaps you struggled to verify that the data was clean because they did not present it well. If there are multiple datasets, some have this trouble | The data is cleaned and tidied.                | The data is spotless and they used tools to visualize the data cleanliness and you were convinced at first glance                      |


# COGS 108 - Data Checkpoint

## Authors

Instructions: REPLACE the contents of this cell with your team list and their contributions. Note that this will change over the course of the checkpoints

This is a modified [CRediT taxonomy of contributions](https://credit.niso.org). For each group member please list how they contributed to this project using these terms:
> Analysis, Background research, Conceptualization, Data curation, Experimental investigation, Methodology, Project administration, Software, Visualization, Writing – original draft, Writing – review & editing

Example team list and credits:
- Alice Anderson: Conceptualization, Data curation, Methodology, Writing - original draft
- Bob Barker:  Analysis, Software, Visualization
- Charlie Chang: Project administration, Software, Writing - review & editing
- Dani Delgado: Analysis, Background research, Visualization, Writing - original draft

## Research Question

Instructions: REPLACE the contents of this cell with your work, including any updates to recover points lost in your proposal feedback



## Background and Prior Work

Instructions: REPLACE the contents of this cell with your work, including any updates to recover points lost in your proposal feedback

## Hypothesis


Instructions: REPLACE the contents of this cell with your work, including any updates to recover points lost in your proposal feedback


## Data

### Data overview

Instructions: REPLACE the contents of this cell with descriptions of your actual datasets.

For each dataset include the following information
- Dataset #1
  - Dataset Name:
  - Link to the dataset:
  - Number of observations:
  - Number of variables:
  - Description of the variables most relevant to this project
  - Descriptions of any shortcomings this dataset has with repsect to the project
- Dataset #2 (if you have more than one!)
  - same as above
- etc

Each dataset deserves either a set of bullet points as above or a few sentences if you prefer that method.

If you plan to use multiple datasets, add a few sentences about how you plan to combine these datasets.

In [ ]:
# Run this code every time when you're actively developing modules in .py files.  It's not needed if you aren't making modules
#
## this code is necessary for making sure that any modules we load are updated here 
## when their source code .py files are modified

%load_ext autoreload
%autoreload 2

In [ ]:
# Setup code -- this only needs to be run once after cloning the repo!
# this code downloads the data from its source to the `data/00-raw/` directory
# if the data hasn't updated you don't need to do this again!

# if you don't already have these packages (you should!) uncomment this line
# %pip install requests tqdm

import sys
sys.path.append('./modules') # this tells python where to look for modules to import

import get_data # this is where we get the function we need to download data

# replace the urls and filenames in this list with your actual datafiles
# yes you can use Google drive share links or whatever
# format is a list of dictionaries; 
# each dict has keys of 
#   'url' where the resource is located
#   'filename' for the local filename where it will be stored 
datafiles = [
    { 'url': 'https://raw.githubusercontent.com/fivethirtyeight/data/refs/heads/master/airline-safety/airline-safety.csv', 'filename':'airline-safety.csv'},
    { 'url': 'https://raw.githubusercontent.com/fivethirtyeight/data/refs/heads/master/bad-drivers/bad-drivers.csv', 'filename':'bad-drivers.csv'}
]

get_data.get_raw(datafiles,destination_directory='data/00-raw/')

### Dataset #1 

Instructions: 
1. Change the header from Dataset #1 to something more descriptive of the dataset
2. Write a few paragraphs about this dataset. Make sure to cover
   1. Describe the important metrics, what units they are in, and giv some sense of what they mean.  For example "Fasting blood glucose in units of mg glucose per deciliter of blood.  Normal values for healthy individuals range from 70 to 100 mg/dL.  Values 100-125 are prediabetic and values >125mg/dL indicate diabetes. Values <70 indicate hypoglycemia. Fasting idicates the patient hasn't eaten in the last 8 hours.  If blood glucose is >250 or <50 at any time (regardless of the time of last meal) the patient's life may be in immediate danger"
   2. If there are any major concerns with the dataset, describe them. For example "Dataset is composed of people who are serious enough about eating healthy that they voluntarily downloaded an app dedicated to tracking their eating patterns. This sample is likely biased because of that self-selection. These people own smartphones and may be healthier and may have more disposable income than the average person.  Those who voluntarily log conscientiously and for long amounts of time are also likely even more interested in health than those who download the app and only log a bit before getting tired of it"
3. Use the cell below to 
    1. load the dataset 
    2. make the dataset tidy or demonstrate that it was already tidy
    3. demonstrate the size of the dataset
    4. find out how much data is missing, where its missing, and if its missing at random or seems to have any systematic relationships in its missingness
    5. find and flag any outliers or suspicious entries
    6. clean the data or demonstrate that it was already clean.  You may choose how to deal with missingness (dropna of fillna... how='any' or 'all') and you should justify your choice in some way
    7. You will load raw data from `data/00-raw/`, you will (optionally) write intermediate stages of your work to `data/01-interim` and you will write the final fully wrangled version of your data to `data/02-processed`
4. Optionally you can also show some summary statistics for variables that you think are important to the project
5. Feel free to add more cells here if that's helpful for you


In [51]:
## YOUR CODE TO LOAD/CLEAN/TIDY/WRANGLE THE DATA GOES HERE
import pandas as pd
import numpy as np

A. The data set has been loaded into a pandas DataFrame called `df`. The first few rows of the DataFrame have been displayed using `df.head()`.

In [74]:
df = pd.read_excel('data/00-raw/early_perceptions_chatgpt.xlsx')
df.head()

Q1   Q2     Q3           Q4  \
0  Afghanistan  2.0   27.0  Afghanistan   
1  Afghanistan  4.0  100.0  Afghanistan   
2  Afghanistan  1.0   21.0      Albania   
3  Afghanistan  1.0   18.0      Ecuador   
4  Afghanistan  2.0   26.0        Spain   

                                                  Q5   Q6   Q7   Q8   Q9  Q10  \
0                                     القدس المفتوحة  1.0  1.0  1.0  1.0  1.0   
1                                                456  2.0  2.0  3.0  2.0  4.0   
2                                               منيا  1.0  1.0  1.0  1.0  1.0   
3  Universidad Indoamerica., Facultad Arquitectur...  2.0  1.0  1.0  1.0  4.0   
4                                De vigo universidad  1.0  2.0  2.0  1.0  2.0   

   ...  Q40g  Q40h  Q40i  Q40j  Q40k  Q40l  Q40m  Q40n  Q40o  source  
0  ...   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN      AR  
1  ...   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN      EN  
2  ...   3.0   2.0   3.0   4.0   3.0   2.0   2.0   1.0   1.0      EN  
3  ...   3.0   3.0   3.0   3.0   3.0   3.0   3.0   3.0   3.0      ES  
4  ...   4.0   3.0   4.0   3.0   3.0   3.0   4.0   3.0   4.0      ES  

[5 rows x 174 columns]

B. The data set is tidy. There is one entry in each row and column space.

C. The shape is 23218 rows and 174 columns.

In [75]:
print('Shape:', df.shape)
print('Column names:', df.columns.tolist())
print(df.info())

Shape: (23218, 174)
Column names: ['Q1', 'Q2', 'Q3', 'Q4', 'Q5', 'Q6', 'Q7', 'Q8', 'Q9', 'Q10', 'Q11', 'Q12', 'Q13', 'Q14', 'Q15', 'Q16', 'Q17', 'Q18a', 'Q18b', 'Q18c', 'Q18d', 'Q18e', 'Q18f', 'Q18g', 'Q18h', 'Q18i', 'Q18j', 'Q18k', 'Q18l', 'Q19a', 'Q19b', 'Q19c', 'Q19d', 'Q19e', 'Q19f', 'Q19g', 'Q19h', 'Q19i', 'Q19j', 'Q20', 'Q21a', 'Q21b', 'Q21c', 'Q21d', 'Q22a', 'Q22b', 'Q22c', 'Q22d', 'Q22e', 'Q22f', 'Q22g', 'Q22h', 'Q22i', 'Q22j', 'Q23a', 'Q23b', 'Q23c', 'Q23d', 'Q24a', 'Q24b', 'Q24c', 'Q24d', 'Q24e', 'Q24f', 'Q24g', 'Q25a', 'Q25b', 'Q25c', 'Q25d', 'Q26a', 'Q26b', 'Q26c', 'Q26d', 'Q26e', 'Q26f', 'Q26g', 'Q26h', 'Q26i', 'Q26j', 'Q27a', 'Q27b', 'Q27c', 'Q27d', 'Q27e', 'Q27f', 'Q27g', 'Q27h', 'Q27i', 'Q27j', 'Q28a', 'Q28b', 'Q28c', 'Q28d', 'Q28e', 'Q28f', 'Q28g', 'Q28h', 'Q28i', 'Q29a', 'Q29b', 'Q29c', 'Q29d', 'Q29e', 'Q29f', 'Q29g', 'Q29h', 'Q29i', 'Q30a', 'Q30b', 'Q30c', 'Q30d', 'Q30e', 'Q30f', 'Q30g', 'Q30h', 'Q30i', 'Q30j', 'Q30k', 'Q30l', 'Q31a', 'Q31b', 'Q31c', 'Q31d', 'Q32a', 

In [76]:
rename_dict = {

    # SECTION 1 – DEMOGRAPHICS
    'Q1': 'citizenship',
    'Q2': 'gender',
    'Q3': 'age',
    'Q4': 'country_study',
    'Q5': 'institution',
    'Q6': 'public_institution',
    'Q7': 'student_status',
    'Q8': 'study_level',
    'Q9': 'first_year',
    'Q10': 'field_of_study',
    'Q11': 'learning_mode',
    'Q12': 'job_confidence',

    # SECTION 2 – USAGE
    'Q13': 'used_chatgpt',
    'Q14': 'chatgpt_version',
    'Q15': 'use_frequency',
    'Q16': 'experience',
    'Q17': 'first_heard',

    # Q18 – TASK USAGE
    'Q18a': 'use_academic',
    'Q18b': 'use_professional',
    'Q18c': 'use_creative',
    'Q18d': 'use_proofreading',
    'Q18e': 'use_brainstorm',
    'Q18f': 'use_translate',
    'Q18g': 'use_summarize',
    'Q18h': 'use_calculate',
    'Q18i': 'use_study',
    'Q18j': 'use_personal',
    'Q18k': 'use_research',
    'Q18l': 'use_coding',

    # SECTION 3 – CAPABILITIES
    'Q19a': 'cap_understand',
    'Q19b': 'cap_respond',
    'Q19c': 'cap_conversation',
    'Q19d': 'cap_efficient_info',
    'Q19e': 'cap_reliable_info',
    'Q19f': 'cap_summarize',
    'Q19g': 'cap_simplify',
    'Q19h': 'cap_traditional_learning',
    'Q19i': 'cap_online_learning',
    'Q19j': 'cap_blended_learning',

    # SECTION 4 – REGULATION
    'Q20': 'institution_policy',
    'Q21a': 'reg_international',
    'Q21b': 'reg_government',
    'Q21c': 'reg_university',
    'Q21d': 'reg_employer',

    # ETHICAL CONCERNS
    'Q22a': 'ethics_unethical',
    'Q22b': 'ethics_cheating',
    'Q22c': 'ethics_plagiarism',
    'Q22d': 'ethics_threat_study',
    'Q22e': 'ethics_mislead',
    'Q22f': 'ethics_privacy',
    'Q22g': 'ethics_reduce_interaction',
    'Q22h': 'ethics_replace_education',
    'Q22i': 'ethics_isolation',
    'Q22j': 'ethics_hinder_learning',

    # ETHICAL RESPONSIBILITY
    'Q23a': 'ethics_consult_prof',
    'Q23b': 'ethics_disclose_use',
    'Q23c': 'ethics_report_misuse',
    'Q23d': 'ethics_protect_info',

    # SECTION 5 – SATISFACTION
    'Q24a': 'sat_vs_google',
    'Q24b': 'sat_vs_professors',
    'Q24c': 'sat_vs_peers',
    'Q24d': 'sat_info_clearer',
    'Q24e': 'sat_assistance',
    'Q24f': 'sat_quality',
    'Q24g': 'sat_accuracy',

    'Q25a': 'att_control',
    'Q25b': 'att_interesting',
    'Q25c': 'att_important',
    'Q25d': 'att_everyday_help',

    # SECTION 6 – ACADEMIC OUTCOMES
    'Q26a': 'acad_access_knowledge',
    'Q26b': 'acad_general_knowledge',
    'Q26c': 'acad_specific_knowledge',
    'Q26d': 'acad_personalized',
    'Q26e': 'acad_efficiency',
    'Q26f': 'acad_motivation',
    'Q26g': 'acad_complete_studies',
    'Q26h': 'acad_engagement',
    'Q26i': 'acad_deadlines',
    'Q26j': 'acad_assignment_quality',

    'Q27a': 'dev_exam_ready',
    'Q27b': 'dev_grades',
    'Q27c': 'dev_outside_class',
    'Q27d': 'dev_internships',
    'Q27e': 'dev_learning_experience',
    'Q27f': 'dev_skills',
    'Q27g': 'dev_personal_growth',
    'Q27h': 'dev_academic_growth',
    'Q27i': 'dev_study_satisfaction',
    'Q27j': 'dev_employability',

    # SECTION 7 – SKILLS
    'Q28a': 'skill_academic_writing',
    'Q28b': 'skill_prof_writing',
    'Q28c': 'skill_typing',
    'Q28d': 'skill_native_language',
    'Q28e': 'skill_foreign_language',
    'Q28f': 'skill_interpersonal',
    'Q28g': 'skill_digital_comm',
    'Q28h': 'skill_info_literacy',
    'Q28i': 'skill_content_creation',

    'Q29a': 'skill_numeracy',
    'Q29b': 'skill_decision',
    'Q29c': 'skill_problem_solving',
    'Q29d': 'skill_analytical',
    'Q29e': 'skill_critical_thinking',
    'Q29f': 'skill_creativity',
    'Q29g': 'skill_data_analysis',
    'Q29h': 'skill_programming',
    'Q29i': 'skill_ai_literacy',

    # SECTION 8 – LABOUR MARKET
    'Q30a': 'labour_reduce_jobs',
    'Q30b': 'labour_new_skills',
    'Q30c': 'labour_edu_market_link',
    'Q30d': 'labour_ai_knowledge',
    'Q30e': 'labour_ai_demand',
    'Q30f': 'labour_inequality',
    'Q30g': 'labour_create_jobs',
    'Q30h': 'labour_change_jobs',
    'Q30i': 'labour_remote_work',
    'Q30j': 'labour_reduce_workload',
    'Q30k': 'labour_productivity',
    'Q30l': 'labour_innovation',

    'Q31a': 'mismatch_skill_shortage',
    'Q31b': 'mismatch_under_skilling',
    'Q31c': 'mismatch_obsolescence',
    'Q31d': 'mismatch_skill_gap',

    # SECTION 9 – EMOTIONS
    'Q32a': 'emotion_bored',
    'Q32b': 'emotion_hopeful',
    'Q32c': 'emotion_sad',
    'Q32d': 'emotion_ashamed',
    'Q32e': 'emotion_calm',
    'Q32f': 'emotion_angry',
    'Q32g': 'emotion_relieved',
    'Q32h': 'emotion_happy',
    'Q32i': 'emotion_proud',
    'Q32j': 'emotion_anxious',
    'Q32k': 'emotion_surprised',
    'Q32l': 'emotion_curious',
    'Q32m': 'emotion_excited',
    'Q32n': 'emotion_confused',
    'Q32o': 'emotion_frustrated',

    # SECTION 10 – PERSONAL
    'Q33': 'study_difficulty',
    'Q34': 'preferred_learning_mode',

    'Q35a': 'study_success',
    'Q35b': 'study_attendance',
    'Q35c': 'study_deadlines',
    'Q35d': 'study_grade_reflect',
    'Q35e': 'study_motivation',
    'Q35f': 'financial_aid_sufficient',
    'Q35g': 'taxes_too_high',
    'Q35h': 'network_employment',
    'Q35i': 'skills_employment',
    'Q35j': 'work_experience',
    'Q35k': 'general_knowledge',
    'Q35l': 'specific_knowledge',
    'Q35m': 'faculty_employability',
    'Q35n': 'want_field_job',

    'Q36': 'area_type',
    'Q37': 'financial_aid',
    'Q38': 'employment_status',
    'Q39': 'economic_status',

    'Q40a': 'person_outgoing',
    'Q40b': 'person_helpful',
    'Q40c': 'person_neat',
    'Q40d': 'person_relaxed',
    'Q40e': 'person_curious',
    'Q40f': 'person_determined',
    'Q40g': 'person_respectful',
    'Q40h': 'person_persistent',
    'Q40i': 'person_secure',
    'Q40j': 'person_artistic',
    'Q40k': 'person_energetic',
    'Q40l': 'person_trusting',
    'Q40m': 'person_dependable',
    'Q40n': 'person_emotionally_stable',
    'Q40o': 'person_original',
}
df = df.rename(columns=rename_dict)
df.head()


citizenship  gender    age country_study  \
0  Afghanistan     2.0   27.0   Afghanistan   
1  Afghanistan     4.0  100.0   Afghanistan   
2  Afghanistan     1.0   21.0       Albania   
3  Afghanistan     1.0   18.0       Ecuador   
4  Afghanistan     2.0   26.0         Spain   

                                         institution  public_institution  \
0                                     القدس المفتوحة                 1.0   
1                                                456                 2.0   
2                                               منيا                 1.0   
3  Universidad Indoamerica., Facultad Arquitectur...                 2.0   
4                                De vigo universidad                 1.0   

   student_status  study_level  first_year  field_of_study  ...  \
0             1.0          1.0         1.0             1.0  ...   
1             2.0          3.0         2.0             4.0  ...   
2             1.0          1.0         1.0             1.0  ...   
3             1.0          1.0         1.0             4.0  ...   
4             2.0          2.0         1.0             2.0  ...   

   person_respectful  person_persistent  person_secure  person_artistic  \
0                NaN                NaN            NaN              NaN   
1                NaN                NaN            NaN              NaN   
2                3.0                2.0            3.0              4.0   
3                3.0                3.0            3.0              3.0   
4                4.0                3.0            4.0              3.0   

   person_energetic  person_trusting  person_dependable  \
0               NaN              NaN                NaN   
1               NaN              NaN                NaN   
2               3.0              2.0                2.0   
3               3.0              3.0                3.0   
4               3.0              3.0                4.0   

   person_emotionally_stable  person_original  source  
0                        NaN              NaN      AR  
1                        NaN              NaN      EN  
2                        1.0              1.0      EN  
3                        3.0              3.0      ES  
4                        3.0              4.0      ES  

[5 rows x 174 columns]

D. Find what data is missing and where.

In [80]:
total_missing = df.isna().sum().sum()
total_cells = df.shape[0] * df.shape[1]

print('Total missing values:', total_missing)
print('Percent missing overall:', round((total_missing / total_cells) * 100, 2), '%')

missing_summary = pd.DataFrame({'missing_count': df.isna().sum(),
'missing_percent': df.isna().mean() * 100
}).sort_values('missing_percent', ascending=False)

missing_summary.head()

Total missing values: 1377068
Percent missing overall: 33.89 %


missing_count  missing_percent
emotion_ashamed           10777        46.416573
emotion_proud             10777        46.416573
emotion_relieved          10777        46.416573
emotion_anxious           10774        46.403652
emotion_excited           10774        46.403652

From this we can see that questions later in the survey have higher missing counts and percentages.

F. Drop rows with missing data from vital columns.

In [73]:
total_missing = df.isna().sum().sum()
total_cells = df.shape[0] * df.shape[1]

print("Total missing values:", total_missing)
print("Percent missing overall:", round((total_missing / total_cells) * 100, 2), "%")

subset_vital = [
    'country_study',
    'use_frequency',
    'use_academic',
    'use_summarize',
    'use_study',
    'use_coding'
]

df = df.dropna(subset=subset_vital)
print('Shape:', df.shape)

Total missing values: 1377068
Percent missing overall: 34.09 %
Shape: (15707, 174)


### Dataset #2 

See instructions above for Dataset #1.  Feel free to keep adding as many more datasets as you need.  Put each new dataset in its own section just like these. 

Lastly if you do have multiple datasets, add another section where you demonstrate how you will join, align, cross-reference or whatever to combine data from the different datasets

Please note that you can always keep adding more datasets in the future if these datasets you turn in for the checkpoint aren't sufficient.  The goal here is demonstrate that you can obtain and wrangle data.  You are not tied down to only use what you turn in right now.

In [ ]:
## YOUR CODE TO LOAD/CLEAN/TIDY/WRANGLE THE DATA GOES HERE


## Ethics

Instructions: REPLACE the contents of this cell with your work, including any updates to recover points lost in your proposal feedback

## Team Expectations 

Instructions: REPLACE the contents of this cell with your work, including any updates to recover points lost in your proposal feedback


## Project Timeline Proposal

Instructions: Replace this with your timeline.  **PLEASE UPDATE your Timeline!** No battle plan survives contact with the enemy, so make sure we understand how your plans have changed.  Also if you have lost points on the previous checkpoint fix them